packages/dds/tree/src/simple-tree/arrayNode.ts

```diff
@@ -61,15 +62,25 @@ export interface ReadonlyArrayNode<out T = TreeNode | TreeLeafValue>
 		Awaited<TreeNode & WithType<string, NodeKind.Array>> {}
 
 /**
- * A generic array type, used to defined types like {@link (TreeArrayNode:interface)}.
+ * A {@link TreeNode} which implements 'readonly T[]' and the array mutation APIs.
  *
- * @privateRemarks
- * Inlining this into TreeArrayNode causes recursive array use to stop compiling.
+ * @typeParam TAllowedTypes - Schema for types which are allowed as members of this array.
+ * @typeParam T - Use Default: Do not specify. Type of values to read from the array.
+ * @typeParam TNew - Use Default: Do not specify. Type of values to write into the array.
+ * @typeParam TMoveFrom - Use Default: Do not specify. Type of node from which children can be moved into this array.
  *
- * @system @sealed @public
+ * @sealed @public
  */
-export interface TreeArrayNodeBase<out T, in TNew, in TMoveFrom = ReadonlyArrayNode>
-	extends ReadonlyArrayNode<T> {
+export interface TreeArrayNode<
+	TAllowedTypes extends Unenforced<ImplicitAllowedTypes> = ImplicitAllowedTypes,
+	out T = [TAllowedTypes] extends [ImplicitAllowedTypes]
```

Comment:
What is the purpose of this check and its two paths? Why was it not necessary before?

Reply:
We don't have an actual extends clause on TAllowedTypes anymore (and can't since its used by the unsafe/recursive types). Therefor `TreeNodeFromImplicitAllowedTypes<TAllowedTypes>` is a compile error if we don't put this conditional in.

Before the recursive types duplicated TreeArrayNode, which was practical to do since they could reuse TreeArrayNodeBase which didn't have an extends clause. Now I don't want to copy this entire APi into the unsafe variant of the types for recursive use, so I tried some workarounds and this worked.

Reply:
Ah, I see. Makes sense, thanks!